openwisp_users/api/filters.py

```diff
@@ -0,0 +1,157 @@
+import swapper
+from django.core.exceptions import ValidationError
+from rest_framework.exceptions import NotFound
+
+Organization = swapper.load_model('openwisp_users', 'Organization')
+
+
+class FilterByOrganization:
+    """
+    Filter queryset based on the access to the organization
+    of the associated model. Use on of the sub-classes
+    """
+
+    def get_queryset(self):
+        qs = super().get_queryset()
+        if self.request.user.is_superuser:
+            return qs
+        return self.get_organization_queryset(qs)
+
+    def get_organization_queryset(self):
+        raise NotImplementedError()
+
+
+class FilterByOrganizationMembership(FilterByOrganization):
+    """
+    Filter queryset by organizations the user is a member of
+    """
+
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=self.request.user.organizations_dict)
+
+
+class FilterByOrganizationManaged(FilterByOrganization):
+    """
+    Filter queryset by organizations managed by user
+    """
+
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=self.request.user.organizations_managed)
+
+
+class FilterByOrganizationOwned(FilterByOrganization):
+    """
+    Filter queryset by organizations owned by user
+    """
+
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=self.request.user.organizations_owned)
+
+
+class FilterByParent:
+    """
+    Filter queryset based on one of the parent objects
+    """
+
+    def get_queryset(self):
+        qs = super().get_queryset()
+        self.assert_parent_exists()
+        return qs
+
+    def assert_parent_exists(self):
+        parent_queryset = self.get_parent_queryset()
+        if not self.request.user.is_superuser:
+            parent_queryset = self.get_organization_queryset(parent_queryset)
+        try:
+            assert parent_queryset.exists()
+        except (AssertionError, ValidationError):
+            raise NotFound(detail='No relevant data found.')
+
+    def get_parent_queryset(self):
+        raise NotImplementedError()
+
+    def get_organization_queryset(self):
+        raise NotImplementedError()
+
+
+class FilterByParentMembership(FilterByParent):
+    """
+    Filter queryset based on parent organization membership
+    """
+
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=self.request.user.organizations_dict)
+
+
+class FilterByParentManaged(FilterByParent):
+    """
+    Filter queryset based on parent organizations managed by user
+    """
+
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=self.request.user.organizations_managed)
+
+
+class FilterByParentOwned(FilterByParent):
+    """
+    Filter queryset based on parent organizations owned by user
+    """
+
+    def get_organization_queryset(self, qs):
+        return qs.filter(organization__in=self.request.user.organizations_owned)
```

Comment:
same concept here, you can eliminate the repetition of these 3 classes as you did below